File: app/src/main/java/com/base/and/base/BaseDialog.java
package com.base.and.base;

import android.app.Activity;
import android.app.Dialog;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import android.view.Gravity;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

import com.base.and.R;
import com.base.and.utils.DensityUtil;

public class BaseDialog<T extends ViewDataBinding> {
    private Dialog mDialog;
    private Activity mContext;
    private T binding;
    private int mWidth;
    private int mGravity;
    private onCreateView onCreateView;

    public BaseDialog(Activity context, int resId, onCreateView<T> onCreateView) {
        this.mContext = context;
        binding = DataBindingUtil.inflate(context.getLayoutInflater(), resId, null, false);
        //设置默认的关闭按钮、取消按钮事件
        View.OnClickListener onClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mDialog.dismiss();
            }
        };
        if (binding.getRoot().getRootView().findViewById(R.id.iv_close) != null) {
            binding.getRoot().getRootView().findViewById(R.id.iv_close).setOnClickListener(onClickListener);
        }
        if (binding.getRoot().getRootView().findViewById(R.id.btn_left) != null) {
            binding.getRoot().getRootView().findViewById(R.id.btn_left).setOnClickListener(onClickListener);
        }
        if (binding.getRoot().getRootView().findViewById(R.id.btn_right) != null) {
            binding.getRoot().getRootView().findViewById(R.id.btn_right).setOnClickListener(onClickListener);
        }
        this.onCreateView = onCreateView;
    }

    public void build() {
        mDialog = new Dialog(mContext, R.style.Theme_AppCompat_Dialog);
        mDialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        Window dialogWindow = mDialog.getWindow();
        dialogWindow.setContentView(binding.getRoot());
        dialogWindow.getDecorView().setPadding(0, 0, 0, 0);
        dialogWindow.setBackgroundDrawable(null);
        //位置
        dialogWindow.setGravity(mGravity == 0 ? Gravity.CENTER : mGravity);
        //设置宽度
        WindowManager.LayoutParams lp = dialogWindow.getAttributes();
        lp.width = mWidth == 0 ? (int) (DensityUtil.getScreenWidth(mContext) * 0.7) : mWidth;
        dialogWindow.setAttributes(lp);
        //外部操作ui 此刻两个参数都有值了
        onCreateView.createView(binding, mDialog);
    }

    public BaseDialog setWidth(int width) {
        this.mWidth = width;
        return this;
    }

    public BaseDialog setGravity(int gravity) {
        this.mGravity = gravity;
        return this;
    }

    public void show() {
        build();
        mDialog.show();
    }

    public interface onCreateView<T> {
        void createView(T binding, Dialog dialog);
    }
}

File: app/src/main/kotlin/com/jeahwan/origin/base/BaseDialog.kt
package com.jeahwan.origin.base

import android.app.Activity
import android.app.Dialog
import android.view.Gravity
import android.view.View
import android.view.Window
import androidx.annotation.StyleRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import com.jeahwan.origin.R
import com.jeahwan.origin.utils.DensityUtil.getScreenWidth
import com.jeahwan.origin.utils.RxTask
import java.util.UUID
import java.util.concurrent.TimeUnit

class BaseDialog<T : ViewDataBinding>(
    private val context: Activity?,
    private val resId: Int,
    private val onCreateView: OnCreateView<T>,
) {
    private var mDialog: Dialog? = null
    private val binding: T? =
        context?.run { DataBindingUtil.inflate(layoutInflater, resId, null, false) }
    private var mWidth = (getScreenWidth() * 0.7).toInt()
    private var mHeight = 0
    private var mGravity = Gravity.CENTER
    private var mWindowAnimationsRes = R.style.dialogBottomEnter // 默认动画

    // 点击dialog外部是否消失
    private var mCancelable = true

    // dialog消失监听
    private var mDismissListener: OnDialogDismissListener? = null

    private var startTime: Long = System.currentTimeMillis()
    private var ignoreTrackEvent: Boolean = false
    private var contentId: String? = null
    private var contentType: String? = null // 1-文章；2-系列课；3-打包课(无)；4-话题；5-作者
    private var pageTile: String = ""

    fun build() {
        if (context == null || binding == null) return
        mDialog =
            Dialog(context, androidx.constraintlayout.widget.R.style.Theme_AppCompat_Dialog).apply {
                val dialogUUID = UUID.randomUUID().toString()
                requestWindowFeature(Window.FEATURE_NO_TITLE)
                setCanceledOnTouchOutside(mCancelable)
                setCancelable(mCancelable)
                window?.run {
                    setContentView(binding.root)
                    decorView.setPadding(0, 0, 0, 0)
                    setBackgroundDrawable(null)
                    //位置
                    setGravity(mGravity)
                //动画
                setWindowAnimations(mWindowAnimationsRes)
                //设置宽度
                attributes = attributes?.apply {
                    width = mWidth
                    if (mHeight != 0) {
                        height = mHeight
                    }
                }
            }
        }
        //外部操作ui 此刻两个参数都有值了
        onCreateView.createView(binding, mDialog!!)
    }

    fun setWidth(width: Int): BaseDialog<*> {
        mWidth = width
        return this
    }

    fun setHeight(height: Int): BaseDialog<*> {
        mHeight = height
        return this
    }

    fun setGravity(gravity: Int): BaseDialog<*> {
        mGravity = gravity
        return this
    }

    fun setCancelable(cancelable: Boolean): BaseDialog<*> {
        mCancelable = cancelable
        return this
    }

    fun setOnDismissListener(listener: OnDialogDismissListener?): BaseDialog<*> {
        mDismissListener = listener
        return this
    }

    fun dismiss() {
        if (mDialog != null && mDialog!!.isShowing) {
            mDialog?.dismiss()
        }
    }

    fun setIgnoreTrackEvent(ignore: Boolean): BaseDialog<*> {
        ignoreTrackEvent = ignore
        return this
    }

    fun setContentId(contentId: String?): BaseDialog<*> {
        this.contentId = contentId
        return this
    }

    fun setContentType(contentType: String?): BaseDialog<*> {
        this.contentType = contentType
        return this
    }

    fun setPageTitle(pageTitle: String): BaseDialog<*> {
        this.pageTile = pageTitle
        return this
    }

    fun show(milliSeconds: Long = 0) {
        RxTask.doInMainThreadDelay({
            if (context != null && !context.isDestroyed) {
                startTime = System.currentTimeMillis()
                build()
                mDialog?.show()
            }
        }, milliSeconds, TimeUnit.MILLISECONDS)
    }

    fun setWindowAnimations(@StyleRes resId: Int): BaseDialog<*> {
        mWindowAnimationsRes = resId
        return this
    }

    interface OnCreateView<T> {
        fun createView(dialogBinding: T, dialog: Dialog)
    }

    interface OnDialogDismissListener {
        fun onDismiss()
    }

    init {
        //设置默认的关闭按钮、取消按钮事件
        binding?.root?.rootView?.findViewById<View?>(R.id.iv_close)?.run {
            setOnClickListener { mDialog?.dismiss() }
        }
    }
}